### PR TITLE
added e2e to successful deployment and test

### PR DESCRIPTION
**Pull request by pascalism**

### CODE REVIEW

File: ext/kubernetes/deployment/init.go [hunk 1]
```diff
@@ -96,6 +97,27 @@ func initiateDeploymentProcedure(d *Deployment) {
 				deploySuccess,
 				strings.Join(d.GetPods(), " "),
 			)
+			if strings.Contains(componentName, "prod") && strings.Contains(clusterName, "eu") {
```
Comment (microwaves) [inline ext/kubernetes/deployment/init.go]: I would suggest to not hardcode this here. Instead, would definitely better to create a configuration entry for conditions or even a template that would give an specific instruction to Signa.

File: ext/kubernetes/deployment/init.go [hunk 2]
```diff
@@ -96,6 +97,27 @@ func initiateDeploymentProcedure(d *Deployment) {
...
+				initiateE2EOnProdEu()
+			}
 		}
 	}()
 }
+
+var initiateE2EOnProdEu = func() ([]byte, error) {
+	cToken := bot.Cfg().CircleCIToken
+	curlPieces := []string{"curl -u ", cToken, ": -d build_parameters[CIRCLE_JOB]=production-build https://circleci.com/api/v1.1/project/github/signavio/pex-e2e-testing/tree/master"}
```
Comment (microwaves) [inline ext/kubernetes/deployment/init.go]: Same applies to this line. Better to not have this hardcoded URL here 😸.

File: ext/kubernetes/deployment/init.go [hunk 3]
```diff
@@ -96,6 +97,27 @@ func initiateDeploymentProcedure(d *Deployment) {
...
+	curlRequest := strings.Join(curlPieces, "")
+	outputCircleCI, error := triggerE2EPipeline(curlRequest)
+	return outputCircleCI, error
+}
+
+var triggerE2EPipeline = func(path string) ([]byte, error) {
+
+	cmd := exec.Command("/bin/sh", "-c", path)
```
Comment (microwaves) [inline ext/kubernetes/deployment/init.go]: Maybe use `/usr/bin/env sh` here.

File: ext/kubernetes/deployment/init_test.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
...0, 34, 32, 58, 32, 102, 97, 108, 115, 101, 44, 10, 32, 32, 34, 106, 111, 98, 95, 110, 97, 109, 101, 34, 32, 58, 32, 110, 117, 108, 108, 44, 10, 32, 32, 34, 115, 116, 97, 114, 116, 95, 116, 105, 109, 101, 34, 32, 58, 32, 110, 117, 108, 108, 44, 10, 32, 32, 34, 99, 97, 110, 99, 101, 108, 101, 114, 34, 32, 58, 32, 110, 117, 108, 108, 44, 10, 32, 32, 34, 112, 108, 97, 116, 102, 111, 114, 109, 34, 32, 58, 32, 34, 50, 46, 48, 34, 44, 10, 32, 32, 34, 111, 117, 116, 99, 111, 109, 101, 34, 32, 58, 32, 110, 117, 108, 108, 44, 10, 32, 32, 34, 118, 99, 115, 95, 117, 114, 108, 34, 32, 58, 32, 34, 104, 116, 116, 112, 115, 58, 47, 47, 103, 105, 116, 104, 117, 98, 46, 99, 111, 109, 47, 115, 105, 103, 110, 97, 118, 105, 111, 47, 112, 101, 120, 45, 101, 50, 101, 45, 116, 101, 115, 116, 105, 110, 103, 34, 44, 10, 32, 32, 34, 97, 117, 116, 104, 111, 114, 95, 110, 97, 109, 101, 34, 32, 58, 32, 110, 117, 108, 108, 44, 10, 32, 32, 34, 110, 111, 100, 101, 34, 32, 58, 32, 110, 117, 108, 108, 44, 10, 32, 32, 34, 99, 97, 110, 99, 101, 108, 101, 100, 34, 32, 58, 32, 102, 97, 108, 115, 101, 44, 10, 32, 32, 34, 97, 117, 116, 104, 111, 114, 95, 101, 109, 97, 105, 108, 34, 32, 58, 32, 110, 117, 108, 108, 10, 125}
```
Comment (microwaves) [inline ext/kubernetes/deployment/init_test.go]: Please, indent this line.

File: ext/kubernetes/deployment/init_test.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+		return b, nil
+	}
+
+	outputCircle, error := triggerE2EPipeline("curl -u $CIRCLECI_Token: -d build_parameters[CIRCLE_JOB]=production-build https://circleci.com/api/v1.1/project/github/signavio/pex-e2e-testing/tree/master")
```
Comment (microwaves) [inline ext/kubernetes/deployment/init_test.go]: Hardcoded URI.

File: ext/kubernetes/deployment/init_test.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+	json.Unmarshal(output, &f)
+	m := f.(map[string]interface{})
+
+	x := false
+
+	for key, value := range m {
+
+		if strings.Compare(key, "reponame") == 0 {
+
+			if strings.Compare(value.(string), "pex-e2e-testing") == 0 {
```
Comment (microwaves) [inline ext/kubernetes/deployment/init_test.go]: Also, hardcoded.

**Comment by microwaves** [comment]
Hi @pascalism, I just added some comments to your code. For documentation purposes, would be great to have a good description of your feature here in this PR. As the project is Open Source, would be also nice to make sure it works for the external users too. Thanks for your contribution!

**Comment by pascalism** [comment]
@microwaves, thanks already for your comments, this is the first prototype for the implementation, next week we'll work on a solution closer to your recommendations. 👍

**Comment by microwaves** [comment]
@pascalism if you need something, just let me know. I can help with the review when the feature is done. Cheers

**Comment by pascalism** [comment]
Changes from Pull Request

The changes I made enable you to add scripts to be executed after a successful deployment. 
By adding an extra Post-Production-Step like this  
```
post-production-step:
   command: foo bar
   cluster: bar foo
```
to the corresponding config file, you can specify the needed request and given circumstances it is supposed to run at.

File: ext/kubernetes/deployment/init_test.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+	json.Unmarshal(output, &f)
+	m := f.(map[string]interface{})
+
+	x := false
+
+	for key, value := range m {
+
+		if key == "reponame" {
+
+			if value.(string) == "pex-e2e-testing" {
```
Comment (jmarten) [inline ext/kubernetes/deployment/init_test.go]: This test does not belong into this repo.
If you like you can create a test, which checks that the triggerRequest function is called when a postDeployment step is defined and the cluster name matches. And that it is not called when no postDeploymentStep is defined or the cluster name does not match.

File: ext/kubernetes/deployment/init_test.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
...0, 34, 32, 58, 32, 102, 97, 108, 115, 101, 44, 10, 32, 32, 34, 106, 111, 98, 95, 110, 97, 109, 101, 34, 32, 58, 32, 110, 117, 108, 108, 44, 10, 32, 32, 34, 115, 116, 97, 114, 116, 95, 116, 105, 109, 101, 34, 32, 58, 32, 110, 117, 108, 108, 44, 10, 32, 32, 34, 99, 97, 110, 99, 101, 108, 101, 114, 34, 32, 58, 32, 110, 117, 108, 108, 44, 10, 32, 32, 34, 112, 108, 97, 116, 102, 111, 114, 109, 34, 32, 58, 32, 34, 50, 46, 48, 34, 44, 10, 32, 32, 34, 111, 117, 116, 99, 111, 109, 101, 34, 32, 58, 32, 110, 117, 108, 108, 44, 10, 32, 32, 34, 118, 99, 115, 95, 117, 114, 108, 34, 32, 58, 32, 34, 104, 116, 116, 112, 115, 58, 47, 47, 103, 105, 116, 104, 117, 98, 46, 99, 111, 109, 47, 115, 105, 103, 110, 97, 118, 105, 111, 47, 112, 101, 120, 45, 101, 50, 101, 45, 116, 101, 115, 116, 105, 110, 103, 34, 44, 10, 32, 32, 34, 97, 117, 116, 104, 111, 114, 95, 110, 97, 109, 101, 34, 32, 58, 32, 110, 117, 108, 108, 44, 10, 32, 32, 34, 110, 111, 100, 101, 34, 32, 58, 32, 110, 117, 108, 108, 44, 10, 32, 32, 34, 99, 97, 110, 99, 101, 108, 101, 100, 34, 32, 58, 32, 102, 97, 108, 115, 101, 44, 10, 32, 32, 34, 97, 117, 116, 104, 111, 114, 95, 101, 109, 97, 105, 108, 34, 32, 58, 32, 110, 117, 108, 108, 10, 125}
```
Comment (microwaves) [inline ext/kubernetes/deployment/init_test.go]: Maybe indent this guy could be a good idea. My Vim would appreciate it! 😸

File: pkg/bot/config.go [hunk 9]
```diff
@@ -48,6 +49,18 @@ type Cluster struct {
 	Kubeconfig string `yaml:"kubeconfig"`
 }
 
+type PostProductionStep struct {
+	Command string `yaml:"command"`
+	Cluster string `yaml:"cluster"`
+}
+
+func (c *Component) HasPostProductionStep() bool {
```
Comment (jmarten) [inline pkg/bot/config.go]: Just realized that postDeploymentStep might be more suitable.

File: ext/kubernetes/deployment/init_test.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+		return byteArrayExample, nil
+	}
+
+	outputCircle, _ := triggerRequest("curl -u SOMETHING")
```
Comment (microwaves) [inline ext/kubernetes/deployment/init_test.go]: Maybe circle offers an URI for tests, you could maybe try something like:

```go
if outputCircle, err := triggerRequest("curl -u <actual-circle-test-uri>"); err != nil {
    outputCircle, err := triggerRequest("curl -u SOMETHING")
    if err != nil {
        return err
    }
}
```

File: pkg/bot/config.go [hunk 11]
```diff
@@ -19,14 +19,15 @@ type Config struct {
...
-	Alias           string   `yaml:"alias"`
+	Name               string `yaml:"name"`
+	Clusters           []Cluster
+	Containers         []Container
+	BootstrapConfig    string             `yaml:"bootstrap-config"`
+	Kubeconfig         string             `yaml:"kubeconfig"`
+	Namespace          string             `yaml:"namespace"`
+	ExecUsers          []string           `yaml:"exec-users"`
+	Alias              string             `yaml:"alias"`
+	PostProductionStep PostProductionStep `yaml:"post-production-step"`
```
Comment (microwaves) [inline pkg/bot/config.go]: 👍

File: pkg/bot/config.go (hunk 9 above)
Comment (microwaves) [reply pkg/bot/config.go]: +1

**Comment by microwaves** [comment]
@pascalism something that I forgot to mention. Could you please create a new Git tag, push and then run `make release` in the root of the repo? After it, please create a release here https://github.com/signavio/signa/releases informing the new feature. I find this mechanism useful to keep track of the changes in the project. Thanks a lot!

**Pull request merged**